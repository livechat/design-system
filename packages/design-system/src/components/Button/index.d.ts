import * as React from 'react';

export type ButtonSize = 'compact' | 'large';

export interface ButtonProps {
  children?: string;
  id?: string;
  primary?: boolean;
  destructive?: boolean;
  disabled?: boolean;
  loading?: boolean;
  size?: ButtonSize;
  outline?: boolean;
  fullWidth?: boolean;
  icon?: any;
  submit?: boolean;
  accessibilityLabel?: string;
  ariaControls?: string;
  ariaExpanded?: boolean;
  onClick?(): void;
  onFocus?(): void;
  onBlur?(): void;
  renderIcon?: string | React.ReactNode;
}
