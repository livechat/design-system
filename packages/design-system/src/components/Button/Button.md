### Primary Button

Use when there is an action that is more important than any other actions and you need to draw attention to it. Don’t use more than one primary button per screen to avoid confusing users. Not every screen needs a primary button.
```js
<Button primary>Primary Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button primary>Primary Button</Button>
  </ComponentHtmlMarkup>
```
### Basic Button

The standard button for most use cases. Only use another style if a button requires more or less visual attention.
```js
  <Button>Basic Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button>Basic Button</Button>
  </ComponentHtmlMarkup>
```
### Destructive Button

Use as a final confirmation for a destructive action such as deleting. These are found mostly in confirmation modals.
```js
  <Button destructive>Destructive Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button destructive>Destructive Button</Button>
  </ComponentHtmlMarkup>
```
### Button Sizes
#### Compact
Use in a table or list where space is tight.
```js
  <Button size="compact" primary>Compact Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button size="compact" primary>Compact Button</Button>
  </ComponentHtmlMarkup>
```
#### Regular
```js
  <Button primary>Regular Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button primary>Regular Button</Button>
  </ComponentHtmlMarkup>
```
#### Large
Use for the main call to action in empty states or for calls to action shown with large illustrations.
```js
<Button size="large" primary>Large Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button size="large" primary>Large Button</Button>
  </ComponentHtmlMarkup>
```
#### Full-width

Button expanded to the full width to fill the parent container (max-width: 320px)
```js
  <Button fullWidth primary>Full-width Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button fullWidth primary>Full-width Button</Button>
  </ComponentHtmlMarkup>
```

### Best practices
 - Prioritize most important action; don’t use more than 1 primary button per screen.
 - Position buttons in consistent locations in the UI.
 - Use short yet descriptive labels ({verb} + {noun}).
 - Use strong verbs to encourage action.
 - Avoid unnecessary words and articles such as the, an, or a.