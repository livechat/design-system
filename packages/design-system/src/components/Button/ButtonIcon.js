import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import getMergedClassNames from '../../utils/getMergedClassNames';
import styles from './style.scss';

const cx = classNames.bind(styles);

export const ButtonIcon = ({ position, children }) => {
  const baseClass = 'btn__icon';
  const classes = getMergedClassNames(
    cx({
      [`${baseClass}`]: true,
      [`${baseClass}-left`]: position === 'left',
      [`${baseClass}-right`]: position === 'right'
    })
  );
  return <i className={classes}>{children}</i>;
};

ButtonIcon.propTypes = {
  position: PropTypes.oneOf(['left', 'right']).isRequired,
  children: PropTypes.node.isRequired
};
