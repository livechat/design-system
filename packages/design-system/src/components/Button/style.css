.btn {
  max-width: 320px;
  position: relative;
  align-items: center;
  justify-content: center;
  user-select: none;
  text-decoration: none;
  transition-property: opacity, border, box-shadow;
  transition-duration: 200ms;
  transition-timing-function: cubic-bezier(0.64, 0, 0.35, 1);
  border-width: 1px;
  border-style: solid;
  text-align: center;
  cursor: pointer;
  border-radius: 4px;
  font-weight: 600;
  display: inline-flex;
  background-color: #fff;
  color: rgba(66,77,87,1);
  border-color: #bcc6d0;
  font-size: 15px;
  padding: 0 16px;
  min-width: 36px;
  min-height: 36px;
  line-height: 36px;
}

.btn:hover {
  opacity: 0.8;
}

.btn:focus {
  outline: 0;
  box-shadow: 0 0 0 1px #427fe1;
}

.btn:disabled {
  opacity: 0.65;
  cursor: not-allowed;
}

.btn .icon {
  color: #4384f5;
  margin-right: 7px;
}

.btn--full-width {
  display: flex;
  width: 100%;
}

.btn--primary {
  background-color: #4384f5;
  color: #fff;
  border-color: #4384f5;
}

.btn--primary:focus {
  box-shadow: 0 0 0 1px #427fe1;
}

.btn--destructive {
  background-color: #f4574c;
  color: #fff;
  border-color: #f4574c;
}

.btn--destructive:focus {
  box-shadow: 0 0 0 1px #f4574c;
}

.btn--large {
  font-size: 15px;
  padding: 0 24px;
  min-width: 42px;
  min-height: 42px;
  line-height: 42px;
}

.btn--compact {
  font-size: 14px;
  padding: 0 16px;
  min-width: 32px;
  min-height: 32px;
  line-height: 32px;
}
