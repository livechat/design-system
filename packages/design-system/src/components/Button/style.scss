@import '../../scss/variables/colors';

.btn {
  position: relative;
  box-sizing: border-box;
  user-select: none;
  text-decoration: none;
  transition-property: opacity, border, color, background-color, box-shadow;
  transition-duration: 200ms;
  transition-timing-function: cubic-bezier(0.64, 0, 0.35, 1);
  border-width: 1px;
  border-style: solid;
  text-align: center;
  cursor: pointer;
  border-radius: 4px;
  font-weight: 600;
  display: inline-block;
  background-color: #fff;
  color: $font-primary-color;
  border-color: $border-primary-color;
  font-size: 15px;
  padding: 7px 16px;
  min-width: 36px;
  min-height: 36px;
  line-height: 1.3;

  &:focus,
  &:hover,
  &:active {
    outline: 0;
  }

  &:hover {
    background-color: $hover-basic-color;
  }

  &:focus {
    box-shadow: 0 0 1px 2px rgba($primary-color, 0.7);
  }

  &:active {
    box-shadow: 0 0 0 1px $border-primary-color;
    transition-duration: 100ms;
  }

  &:disabled {
    cursor: not-allowed;
  }

  &--disabled {
    opacity: 0.65;
  }

  &--full-width {
    width: 100%;
  }

  &--primary {
    background-color: $primary-color;
    color: #fff;
    border-color: $primary-color;

    &:hover {
      background-color: $hover-primary-color;
    }

    &:active {
      box-shadow: 0 0 0 1px $primary-color;
    }
  }

  &--destructive {
    background-color: $error-color;
    color: #fff;
    border-color: $error-color;

    &:hover {
      background-color: $hover-red-color;
    }

    &:active {
      box-shadow: 0 0 0 1px $error-color;
    }
  }

  &--large {
    padding: 10px 24px;
    min-width: 42px;
    min-height: 42px;
  }

  &--compact {
    padding: 5px 16px;
    min-width: 32px;
    min-height: 32px;
  }

  .icon {
    color: $primary-color;
    margin-right: 7px;
  }

  &--loading {
    background-color: $loading-btn-basic-color;
    border-color: #d0d2d5;
    color: transparent;
    transition: all 0.2s cubic-bezier(0.64, 0, 0.35, 1);

    &.btn--destructive {
      background-color: $loading-btn-error-color;
      border-color: transparent;
    }

    &.btn--primary {
      background-color: $loading-btn-primary-color;
      border-color: transparent;
    }
  }

  .btn__loader {
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
    left: 50%;
  }
  .btn__content {
    display: inline-block;
  }
}
