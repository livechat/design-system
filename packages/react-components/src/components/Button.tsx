import * as React from 'react';
import cx from 'classnames';
import { Loader } from './Loader';

export type ButtonSize = 'compact' | 'medium' | 'large';

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  kind?: 'basic' | 'primary' | 'secondary' | 'destructive' | 'text';
  size?: ButtonSize;
  disabled?: boolean;
  loading?: boolean;
  fullWidth?: boolean;
  loaderLabel?: string;
  icon?: React.ReactElement;
  iconPosition?: 'left' | 'right';
}

const baseClass = 'lc-btn';

export const Button: React.FC<ButtonProps> = ({
  loading = false,
  disabled = false,
  type = 'button',
  fullWidth = false,
  kind = 'basic',
  size = 'medium',
  icon = null,
  iconPosition = 'left',
  loaderLabel,
  className,
  children,
  ...props
}) => {
  const isDisabled = loading || disabled;

  const mergedClassNames = cx(
    className,
    baseClass,
    `${baseClass}--${kind}`,
    `${baseClass}--${size}`,
    {
      [`${baseClass}--loading`]: loading,
      [`${baseClass}--full-width`]: fullWidth,
      [`${baseClass}--icon-only`]: !children && icon,
    }
  );

  return (
    <button
      className={mergedClassNames}
      disabled={isDisabled}
      type={type}
      {...props}
    >
      {loading && (
        <Loader
          size="small"
          label={loaderLabel}
          className={`${baseClass}__loader`}
        />
      )}
      {icon &&
        React.cloneElement(icon, {
          className: cx(
            `${baseClass}__icon`,
            `${baseClass}__icon--${iconPosition}`
          ),
          disabled,
        })}
      <div className={`${baseClass}__content`}>{children}</div>
    </button>
  );
};
