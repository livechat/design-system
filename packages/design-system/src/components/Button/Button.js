import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './style.scss';
import getMergedClassNames from '../../utils/getMergedClassNames';
import { Loader } from '../Loader';
import { noticeAboutDeprecation } from '../../helpers/notice-about-deprecation';
import { ButtonIcon } from './ButtonIcon';

const cx = classNames.bind(styles);
const acceptedSizes = ['large', 'compact'];
const acceptedKinds = ['primary', 'secondary', 'destructive', 'text'];

const getDeprecatedKind = (primary, destructive, secondary) => {
  if (primary) {
    return 'primary';
  }

  if (destructive) {
    return 'destructive';
  }

  if (secondary) {
    return 'secondary';
  }

  return null;
};

const Button = React.forwardRef((props, ref) => {
  const {
    children,
    primary,
    destructive,
    disabled,
    loading,
    loaderLabel,
    icon,
    size,
    fullWidth,
    submit,
    secondary,
    accessibilityLabel,
    ariaControls,
    ariaExpanded,
    className,
    type: htmlType,
    kind,
    iconPosition,
    ...buttonProps
  } = props;

  const isDisabled = disabled || loading;
  const type = submit ? 'submit' : htmlType || 'button';
  const isValidKind = kind && acceptedKinds.includes(kind);
  const deprecatedKind = getDeprecatedKind(primary, destructive, secondary);
  const buttonKind = (isValidKind && kind) || deprecatedKind || null;

  if (deprecatedKind) {
    noticeAboutDeprecation(
      `deprecated prop '${deprecatedKind}' in Button component - please use 'kind' prop instead`
    );
  }

  const baseClass = 'btn';
  const mergedClassNames = getMergedClassNames(
    cx({
      [baseClass]: true,
      [`${baseClass}--disabled`]: disabled,
      [`${baseClass}--loading`]: loading,
      [`${baseClass}--full-width`]: fullWidth,
      [`${baseClass}--${buttonKind}`]: !!buttonKind,
      [`${baseClass}--${size}`]: acceptedSizes.some(s => s === size)
    }),
    className
  );

  return (
    <button
      {...buttonProps}
      ref={ref}
      className={mergedClassNames}
      type={type}
      disabled={isDisabled}
      role={loading ? 'alert' : undefined}
      aria-busy={loading ? true : undefined}
      aria-label={accessibilityLabel}
      aria-controls={ariaControls}
      aria-expanded={ariaExpanded}
    >
      {loading && (
        <Loader
          size="small"
          label={loaderLabel}
          className={styles[`${baseClass}__loader`]}
          labelClassName={styles[`${baseClass}__loader-label`]}
        />
      )}
      {icon && <ButtonIcon position={iconPosition}>{icon}</ButtonIcon>}
      {children && <div>{children}</div>}
    </button>
  );
});

Button.defaultProps = {
  iconPosition: 'left'
};

Button.propTypes = {
  accessibilityLabel: PropTypes.string,
  ariaControls: PropTypes.string,
  ariaExpanded: PropTypes.string,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  id: PropTypes.string,
  disabled: PropTypes.bool,
  /**
   * Sets button width to max-width=320px
   */
  fullWidth: PropTypes.bool,
  icon: PropTypes.node,
  loading: PropTypes.bool,
  loaderLabel: PropTypes.node,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  /**
   * Size of button
   */
  size: PropTypes.oneOf(['compact', 'large']),
  submit: PropTypes.bool,
  type: PropTypes.string,
  /**
   * Type of button
   */
  kind: PropTypes.oneOf(['primary', 'secondary', 'destructive', 'text']),
  /**
   * Position of provided icon
   */
  iconPosition: PropTypes.oneOf(['left', 'right']),
  /**
   * DEPRECATED - use 'kind' instead
   */
  primary: PropTypes.bool,
  /**
   * DEPRECATED - use 'kind' instead
   */
  destructive: PropTypes.bool,
  /**
   * DEPRECATED - use 'kind' instead
   */
  secondary: PropTypes.bool
};

export default Button;
