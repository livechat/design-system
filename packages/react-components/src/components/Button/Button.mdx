import { Canvas, ArgsTable, Meta, Title } from '@storybook/blocks';
import { Figma } from '@storybook/addon-designs/blocks';
import * as ButtonStories from './Button.stories';
import {
  BlockColumnsContainer,
  BlockColumns,
} from '../../stories/components/BlockColumns/BlockColumns';

<Meta of={ButtonStories} />

<Title>Button</Title>

[Intro](#Intro) | [Component API](#ComponentAPI) | [Content Spec](#ContentSpec)

## Intro <a id="Intro" />

Button used to initialise user actions. Button labels clearly express the action that will occur when the user interacts with them.

<Canvas of={ButtonStories.Default} sourceState="none" />

#### Example implementation

```jsx
<Button kind="primary" size="large" onClick={() => {}}>
  Button text
</Button>
```

<BlockColumnsContainer>
  <BlockColumns label="When to Use">
    - Use buttons when you want users to perform an action or trigger a specific
    function. - Buttons are suitable for primary call-to-action (CTA) elements,
    secondary actions, or supporting functions.
  </BlockColumns>
  <BlockColumns isNegative label="When Not to Use">
    - Avoid using buttons when the action is purely informational and doesn't
    require user interaction. - Don't use buttons for non-actionable elements or
    decorative purposes.
  </BlockColumns>
</BlockColumnsContainer>

## Component API <a id="ComponentAPI" />

<ArgsTable />

## Content Spec <a id="ContentSpec" />

<Figma
  showLink
  height="100%"
  url="https://www.figma.com/file/9FDwjR8lYvincseDkKypC4/Component-Documentations?type=design&node-id=613-36946&mode=dev"
/>
