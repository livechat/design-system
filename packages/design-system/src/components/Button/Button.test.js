import * as React from 'react';
import { shallow } from 'enzyme';
import Button from './Button';

describe('Button', () => {
  const shallowButton = (props = {}) =>
    shallow(<Button {...props}>Click me</Button>);

  it('should render any type of button', () => {
    const primaryButton = shallowButton({
      kind: 'primary'
    });
    const destructiveButton = shallowButton({
      kind: 'destructive'
    });
    const secondaryButton = shallowButton({
      kind: 'secondary'
    });
    const textButton = shallowButton({
      kind: 'text'
    });
    const basicButton = shallowButton();

    expect(primaryButton).toMatchSnapshot();
    expect(destructiveButton).toMatchSnapshot();
    expect(secondaryButton).toMatchSnapshot();
    expect(textButton).toMatchSnapshot();
    expect(basicButton).toMatchSnapshot();
  });

  it('should render any size of button', () => {
    const compactButton = shallowButton({ size: 'compact' });
    const largeButton = shallowButton({ size: 'large' });

    expect(compactButton).toMatchSnapshot();
    expect(largeButton).toMatchSnapshot();
  });

  it('should render loading button', () => {
    const loadingButton = shallowButton({
      loading: true,
      loaderLabel: 'Processing'
    });

    expect(loadingButton).toMatchSnapshot();
  });

  it('should render full-width button', () => {
    const fullWidthButton = shallowButton({ fullWidth: true });

    expect(fullWidthButton).toMatchSnapshot();
  });

  it('should render disabled button', () => {
    const disabledButton = shallowButton({ disabled: true });

    expect(disabledButton).toMatchSnapshot();
  });

  it('on click, focus and blur should call passed in props functions', () => {
    const clickFn = jest.fn();
    const focusFn = jest.fn();
    const blurFn = jest.fn();

    const component = shallow(
      <Button onClick={clickFn} onFocus={focusFn} onBlur={blurFn}>
        Click me
      </Button>
    );
    const button = component.find('button');

    button.simulate('click');
    expect(clickFn).toHaveBeenCalled();

    button.simulate('focus');
    expect(focusFn).toHaveBeenCalled();

    button.simulate('blur');
    expect(blurFn).toHaveBeenCalled();
  });
});
