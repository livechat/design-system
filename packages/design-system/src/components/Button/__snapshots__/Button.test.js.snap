// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render any size of button 1`] = `
<button
  className="btn btn--compact"
  type="button"
>
  <div>
    Click me
  </div>
</button>
`;

exports[`Button should render any size of button 2`] = `
<button
  className="btn btn--large"
  type="button"
>
  <div>
    Click me
  </div>
</button>
`;

exports[`Button should render any type of button 1`] = `
<button
  className="btn btn--primary"
  type="button"
>
  <div>
    Click me
  </div>
</button>
`;

exports[`Button should render any type of button 2`] = `
<button
  className="btn btn--destructive"
  type="button"
>
  <div>
    Click me
  </div>
</button>
`;

exports[`Button should render any type of button 3`] = `
<button
  className="btn btn--secondary"
  type="button"
>
  <div>
    Click me
  </div>
</button>
`;

exports[`Button should render any type of button 4`] = `
<button
  className="btn btn--text"
  type="button"
>
  <div>
    Click me
  </div>
</button>
`;

exports[`Button should render any type of button 5`] = `
<button
  className="btn"
  type="button"
>
  <div>
    Click me
  </div>
</button>
`;

exports[`Button should render disabled button 1`] = `
<button
  className="btn btn--disabled"
  disabled={true}
  type="button"
>
  <div>
    Click me
  </div>
</button>
`;

exports[`Button should render full-width button 1`] = `
<button
  className="btn btn--full-width"
  type="button"
>
  <div>
    Click me
  </div>
</button>
`;

exports[`Button should render loading button 1`] = `
<button
  aria-busy={true}
  className="btn btn--loading"
  disabled={true}
  role="alert"
  type="button"
>
  <Loader
    className="btn__loader"
    label="Processing"
    labelClassName="btn__loader-label"
    size="small"
  />
  <div>
    Click me
  </div>
</button>
`;
